src/memory/spi_mem.c

```diff
@@ -0,0 +1,618 @@
+        return NULL;
+    }
+
+    uint8_t* buffer = malloc(size + 4);
+    if (!buffer) {
+        util_log("Memory allocation failed");
+        return NULL;
+    }
+
+    _spi_mem_read(address, size, buffer);
+
+    // shift the read data at the beginning of the buffer, overriding the command and the address.
+    for (int i = 0; i < (int)size; i++) {
+        buffer[i] = buffer[i + 4];
+    }
+    return buffer;
+}
+
+static bool _spi_mem_page_write(uint32_t page_addr, const uint8_t* input)
+{
+    if (page_addr % SPI_MEM_PAGE_SIZE != 0) {
+        util_log("Invalid page write address %p", (void*)(uintptr_t)page_addr);
+        return false;
+    }
+
+    uint8_t buffer[SPI_MEM_PAGE_SIZE + 4];
+    // --- Enable Write ---
+    buffer[0] = CMD_WREN;
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 1);
+    _spi_mem_cs_high();
+
+    // --- Page Program (write 4 bytes) ---
+    buffer[0] = CMD_PP;
+    buffer[1] = (page_addr >> 16) & 0xFF;
+    buffer[2] = (page_addr >> 8) & 0xFF;
+    buffer[3] = page_addr & 0xFF;
+    memcpy(&buffer[4], input, SPI_MEM_PAGE_SIZE);
+
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 4 + SPI_MEM_PAGE_SIZE);
+    _spi_mem_cs_high();
+
+    // --- Wait for write to end ---
+    _spi_mem_wait();
+
+    return true;
+}
+
+bool spi_mem_write(uint32_t address, const uint8_t* input, size_t size)
+{
+    if (address + size - 1 > MEMORY_LIMIT || size < 1) {
+        util_log("Invalid write address %p or size %i", (void*)(uintptr_t)address, (int)size);
+        return false;
+    }
+
+    uint32_t initial_sector_addr = address & SECTOR_MASK;
+    uint32_t final_sector_addr = ((address + size - 1) & SECTOR_MASK) + SPI_MEM_SECTOR_SIZE;
+    uint16_t sectors = (final_sector_addr - initial_sector_addr) / SPI_MEM_SECTOR_SIZE;
+
+    // read all the affected sectors data
+    uint8_t* buffer = spi_mem_read(initial_sector_addr, (size_t)(sectors * SPI_MEM_SECTOR_SIZE));
+    if (!buffer) {
+        return false;
+    }
+
+    // update data in the buffer
+    memcpy(&buffer[address - initial_sector_addr], input, size);
+
+    // erase sectors and write data
+    for (int i = 0; i < sectors; i++) {
+        uint32_t sector_addr = initial_sector_addr + (i * SPI_MEM_SECTOR_SIZE);
+        if (!spi_mem_sector_erase(sector_addr)) {
+            free(buffer);
+            return false;
+        }
+        for (int p = 0; p < (SPI_MEM_SECTOR_SIZE / SPI_MEM_PAGE_SIZE); p++) {
+            uint32_t page_addr = sector_addr + p * SPI_MEM_PAGE_SIZE;
+            if (!_spi_mem_page_write(
+                    page_addr, &buffer[(i * SPI_MEM_SECTOR_SIZE) + (p * SPI_MEM_PAGE_SIZE)])) {
+                free(buffer);
+                return false;
+            }
+        }
+    }
+    free(buffer);
+    return true;
+}
+
+bool spi_mem_verify_erased(void)
+{
+    uint8_t buffer[SPI_MEM_BLOCK_SIZE + 4];
+    for (int i = 0; i < (SPI_MEM_MEMORY_SIZE / SPI_MEM_BLOCK_SIZE); i++) {
```

Comment:
why not `uint32_t i`? you cast it to that one line below anyway.

In all the other loops where you use int, `uint32_t` might be clearer too?